gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobId.java

```diff
@@ -32,51 +32,87 @@
   private static final long serialVersionUID = -6156002883225601925L;
   private final String bucket;
   private final String name;
+  private final Long generation;
 
-  private BlobId(String bucket, String name) {
+  private BlobId(String bucket, String name, Long generation) {
     this.bucket = bucket;
     this.name = name;
+    this.generation = generation;
   }
 
+  /**
+   * Returns the name of the containing bucket.
+   */
   public String bucket() {
     return bucket;
   }
 
+  /**
+   * Returns the blob's name.
+   */
   public String name() {
     return name;
   }
 
+  /**
+   * Returns blob's data generation. Used for versioning.
+   */
+  public Long generation() {
+    return generation;
+  }
+
   @Override
   public String toString() {
     return MoreObjects.toStringHelper(this)
         .add("bucket", bucket())
         .add("name", name())
+        .add("generation", generation())
         .toString();
   }
 
   @Override
   public int hashCode() {
-    return Objects.hash(bucket, name);
+    return Objects.hash(bucket, name, generation);
   }
 
   @Override
   public boolean equals(Object obj) {
     return obj instanceof BlobId && Objects.equals(bucket, ((BlobId) obj).bucket)
-        && Objects.equals(name, ((BlobId) obj).name);
+        && Objects.equals(name, ((BlobId) obj).name)
+        && Objects.equals(generation, ((BlobId) obj).generation);
   }
 
   StorageObject toPb() {
     StorageObject storageObject = new StorageObject();
     storageObject.setBucket(bucket);
     storageObject.setName(name);
+    storageObject.setGeneration(generation);
     return storageObject;
   }
 
+  /**
+   * Creates a {@code BlobId} object. Generation is set to {@code null}.
+   *
+   * @param bucket name of the containing bucket
+   * @param name blob's name
+   */
   public static BlobId of(String bucket, String name) {
-    return new BlobId(checkNotNull(bucket), checkNotNull(name));
+    return new BlobId(checkNotNull(bucket), checkNotNull(name), null);
+  }
+
+  /**
+   * Creates a {@code BlobId} object.
+   *
+   * @param bucket name of the containing bucket
+   * @param name blob's name
+   * @param generation blob's data generation, used for versioning
```

Comment:
It would be good to mention here (and/or maybe in the class-level javadoc) that the API treats `null` generation as latest.